### DISCUSSION TITLE
Crossfade on skip track

### DISCUSSION THREAD

**Discussion by agentgonzo** [MusicPlayerDaemon/MPD#1849] [Q&A] [answered]
How does crossfade work with MPD when you skip track? I am doing a tinkering project and want to trigger changing the music to a given track when an event occurs, but it needs to be a crossfade, rather than an immediate jump.  Is this possible? If not, does anyone know how easy it might be to modify the code do make it crossfade when skipping to a track?

**Comment by MaxKellermann** [comment, marked as answer]
This feature you describe has been requested many times over the years (e.g. https://github.com/MusicPlayerDaemon/MPD/issues/1821), but has never been implemented. I do see that it's an important feature, but every time I take time for MPD, more important things are lurking for my attention.
Implementing this is not simple, but doable, for somebody with enough motivation. If you want to get your hands dirty, I'll be happy to give advice.

**Comment by agentgonzo** [comment]
Bother. Thanks Max. It's a shame. If it wasn't too much effort, I would probably do it, but if it's not simple then I think it'll probably be too much for me to do. I haven't done C code in donkeys years. I could probably hack it for my own purposes (with guidance) but to make it decent enough for use by other people I think would be too much work.

**Comment by agentgonzo** [comment]
Actually, could you give me some advice on where to look to see if I can get something implemented?

**Reply by MaxKellermann** [reply]
https://github.com/MusicPlayerDaemon/MPD/blob/master/src/player/Thread.cxx
This is the heart of the MPD player engine; here data from the decoder gets collected, sometimes the chunks of two songs get mixed together (for cross-fading), and everything is submitted to the output(s).
The chunks of audio are defined here: https://github.com/MusicPlayerDaemon/MPD/blob/master/src/MusicChunk.hxx
The MusicPipe is a queue of chunks; the decoder fills a the queue while the player consumes it, and the player fills the next queue that leads to the outputs: https://github.com/MusicPlayerDaemon/MPD/blob/master/src/MusicPipe.hxx

**Reply by agentgonzo** [reply]
So presumably,  the player will fill the pipe with chunks from the current/next song as required. Does it fill a full track's worth of chunks at a time, or just a few chuckns from a track?

So if we want to do crossfade on skip, then we could change the skip behaviour to something like
* clear the pipe
* move player reference/pointer straight to the start of the next track
* fill the pipe with chunks from the beginning next track (rather than the current position in the current track)?

**Reply by MaxKellermann** [reply]
Not a full track - the chunks are allocated from a fixed-size buffer (the "MusicBuffer" which iirc defaults to 2 MB). If the MusicBuffer is full, the decoder waits until a chunk is returned (because it has been played by the output) before it continues decoding more of the song.
With today's computers, you could easily allocate a gigabyte or so and keep many songs decoded in memory, but while that sounds like an improvement, it actually wastes a lot of resources for no real advantage. It's good that there's an upper limit, and 2 MB is generous enough.
Note that you can't just clear the pipe - the producer cannot consume anything! It should not interfere with the consume-side of the pipe. And if you do clear the pipe, the part that was already in the pipe is going to be skipped - it will be missing during playback, and the user won't like that. I suggest not clearing the pipe; just keep on playing what was already scheduled to be player - there's an upper limit of 64 chunks between player and outputs (see the first line in `PlayNextChunk()`) - and only those 64 chunks will continue playing without cross-fading, and after that, you can begin cross-fading.
For actual cross-fading, you take a chunk from the old AND the new song, link them together with the "MusicChunk.other" field and set the "mix_ratio" field; each successive double-chunk will have a different mix_ratio.
